src/snooping.jl

```diff
@@ -1,23 +1,44 @@
 using Pkg, Serialization
 
 
-function snoop(package, tomlpath, snoopfile, outputfile, reuse = false, blacklist = [])
+const packages_needing_initialization = Symbol[]
 
-    command = """
-    using Pkg, PackageCompiler
-    """
+"""
+    init_package!(packages::Symbol...)
+Globally add `packages` to the list of packages that need their `__init__`
+method called when ahead of time compiling.
 
-    if tomlpath !== nothing
-        command *= """
-        empty!(Base.LOAD_PATH)
-        # Take LOAD_PATH from parent process
-        append!(Base.LOAD_PATH, $(repr(Base.LOAD_PATH)))
-        Pkg.activate($(repr(tomlpath)))
-        Pkg.instantiate()
-        """
-    end
+(adds it only for the current session).
+"""
+function init_package!(packages::Symbol...)
+    push!(packages_needing_initialization, packages...)
+    # TODO use set?!
+    unique!(packages_needing_initialization)
```

Comment:
I'm not sure how `init_package!` would be used, but it would be nice if it can track the order that the packages are initialized (so `Set` is out).  I think it is required when the top-level code of downstream packages needs global states configured by `__init__` of the upstream packages.  See: #192

Reply:
good point! It seems like you wrote a method to sort packages topological, would you mind adding that? Although, in this PR I'm not initializing packages by default anymore... You need to do it explicitly by calling `init_package!(mypackages...)` so in theory the user could do that. If we add a toposort method, we could protect against user error though ;)

Reply:
I just found an internal function in Pkg that I can use so I didn't need to write anything complicated.  But my code has some problems:

* This internal function is changed between Julia 1.0 and 1.1.  So my code won't work as-is for Julia 1.0.  (But ultimately we can just "vendor" the related code from Pkg so it's an easy problem.)
* The hard part is that being in Manifest.toml does not necessary mean that it should be in the system image.

    * Project.toml/REQUIRE info may be old.  Packages that work only for Julia 0.6 may be listed as dependency even though they are not imported anymore.
    * Some packages are appropriate only for some OSes.  They are usually conditionally imported by something like `if Sys.isapple(); using ...; end`.  This is no good way to avoid importing those packages.

> not initializing packages by default anymore

This is a hard problem but I think a slightly less worse option for now is to initialize all packages by default.  IIUC, we need to be able to evaluate the all top-level statements in the `--output-o` mode, right?  Those top-level statements may require the runtime of the upstream packages.  For example, it's conceivable that someone wants to compute a constant or implement a macro using PyCall functions.  However, PyCall needs to setup Python runtime in its `__init__`.

I think it would be nice if Julia in `--output-o` mode uses precompilation cache.  This way, we don't need to evaluate any code so we can forget about the `__init__` issue.  I bring up this in https://discourse.julialang.org/t/22956/49 as well.  I'm wondering if it is possible to get more attention to this issue...